libcudacxx/include/cuda/std/__type_traits/enable_if.h

```diff
@@ -32,12 +32,33 @@ struct _CCCL_TYPE_VISIBILITY_DEFAULT enable_if<true, _Tp>
 };
 
 template <bool _Bp, class _Tp = void>
-using __enable_if_t _LIBCUDACXX_NODEBUG_TYPE = typename enable_if<_Bp, _Tp>::type;
+using enable_if_t _CCCL_NODEBUG_ALIAS = typename enable_if<_Bp, _Tp>::type;
+
+#if defined(_CCCL_COMPILER_NVHPC) || defined(_CCCL_COMPILER_ICC) \
+  || (defined(_CCCL_COMPILER_GCC) && _CCCL_GCC_VERSION < 80000)
+
+template <bool _Bp, class _Tp = void>
+using __enable_if_t _CCCL_NODEBUG_ALIAS = typename enable_if<_Bp, _Tp>::type;
+
+#else // ^^^ _CCCL_COMPILER_NVHPC || _CCCL_COMPILER_ICC ^^^ / vvv !_CCCL_COMPILER_NVHPC && !_CCCL_COMPILER_ICC vvv
+
+// This is an optimization of the common `enable_if_t` alias. Instead of instantiating at least one class per
+// specialization we instantate exactly two
+template <bool>
+struct __enable_if_impl
+{};
+
+template <>
+struct __enable_if_impl<true>
+{
+  template <class _Tp>
+  using type _CCCL_NODEBUG_ALIAS = _Tp;
+};
 
-#if _CCCL_STD_VER > 2011
 template <bool _Bp, class _Tp = void>
-using enable_if_t = typename enable_if<_Bp, _Tp>::type;
-#endif
+using __enable_if_t _CCCL_NODEBUG_ALIAS = typename __enable_if_impl<_Bp>::template type<_Tp>;
+
+#endif // !_CCCL_COMPILER_NVHPC && !_CCCL_COMPILER_ICC
```

Comment:
can't we just make `enable_if_t` do the not-stupid thing?

Reply:
Its explicitly defined in the standard that way, so it might be potentially observable

Reply:
That said, I wouldnt really mind.

If someone tries to be clever let them have it

Reply:
@griwes any objections?

Reply:
I did replace `enable_if_t` with the efficient implementation when possible

Reply:
Had to revert that too